### DISCUSSION TITLE
Testing and Performance Testing

### DISCUSSION THREAD

**Discussion by adrianmichalke** [nebulastream/nebulastream#3] [Ideas]
# Description
In this Discussion, we want to find out how to best test and performance test NebulaStream.

**Comment by adrianmichalke** [comment]
### Make NebulaStream Profileable

**Comment by ls-1801** [comment]
### Sanitized/Hardend Builds

We want to distribute a Debug build version with Sanitizers and general binary Hardening techniques enabled to catch runtime errors, like memory leaks, corruption, overflows, UB, data races,...

**Comment by klauck** [comment]
### Testing Framework

I assume we want to stick to GoogleTest but what are other things we want to keep or change:

- Tests are organized (in a folder) per module
- De-duplicate test code

**Reply by ls-1801** [reply]
First of all I would want to identify well tested and poorly tested components in the system using code coverage.
I am not sure if this would also allow us to find redundant tests.

Secondly I would want to identify long running unit tests, which in my opinion are bad unit tests.

Lastly running unit tests includes:
- running in debug with all possible asserts enabled
- enabling all sanitizers and hardening we can get our hands on.
